File: 0907-koko-eating-bananas/0907-koko-eating-bananas.java
class Solution {
    private boolean f(int[] piles, int BananasPerHour, int h){
        int timeToComplete = 0;
        for(int i : piles){
            timeToComplete += ((i + BananasPerHour - 1) / BananasPerHour);
            if(timeToComplete > h){
                return false;
            }
        }
        return true;
    }
    public int minEatingSpeed(int[] piles, int h) {
        int low = 1, high = Arrays.stream(piles).max().getAsInt(), ans = -1;
        while(low <= high){
            int mid = low + (high - low)/2;
            if(f(piles, mid, h)){
                ans = mid;
                high = mid - 1;
            } else{
                low = mid + 1;
            }
        }

        return ans;
    }
}

File: 0907-koko-eating-bananas/0907-koko-eating-bananas.cpp
class Solution {
private:
    int fun(vector<int>& piles, int h, int val){
        int cnt = 0;
        for(int i : piles){
            cnt += ceil((double)i / (double)val);
        }
        return cnt <= h;
    }

public:
    int minEatingSpeed(vector<int>& piles, int h) {
        int l = 1, r = INT_MAX, ans = -1;
        while(l <= r){
            long m = l + (r - l) / 2;
            if(fun(piles, h, m)){
                ans = m; r = m - 1;
            } else{
                l = m + 1;
            }
        }

        return ans;
    }
};